### DISCUSSION TITLE
models/yolov5/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#10054] [Docs]
# models/yolov5/

Discover YOLOv5u, a boosted version of the YOLOv5 model featuring an improved accuracy-speed tradeoff and numerous pre-trained models for various object detection tasks.

https://docs.ultralytics.com/models/yolov5/

**Comment by MikolajBaranski** [comment]
My questions (as of 12th April 2024) is whether there is any documentation regarding the Yolov5u model. The description on this page states that the Yolov5u incorporates the objectness-free split head into the Yolov5 architecture. Is there any further documentation on this model? The Yolov5 GitHub doesn't seem to present this model anywhere. While the page of v7.0 update only describes the introduction of segmentation. Is there any resource going into more detail into this specific release of the model?

I see there is also documentation regarding the architecture of the Yolov5 v6.1/6.0 model (link below for reference). Is the Yolov5u an adaptations of that architecture? If so which part is changed?

Github: https://github.com/ultralytics/yolov5
Architecture: https://docs.ultralytics.com/yolov5/tutorials/architecture_description/

**Reply by pderrenger** [reply]
Hey there! 😊 It looks like you're exploring the details of our YOLOv5u model. Indeed, YOLOv5u builds upon the YOLOv5 architecture, especially by incorporating an anchor-free, objectness-free split head, a feature inspired by advancements in YOLOv8. This enhancement is aimed at refining the model's ability to detect objects with a better accuracy-speed tradeoff.

Although it seems there might not be detailed standalone documentation on the YOLOv5u model in the Ultralytics GitHub repo, the key updates and features you’re interested in are described in our [Ultralytics YOLOv5 Overview](https://docs.ultralytics.com/models/yolov5/). This page highlights the integration of the anchor-free split head into the model and its impact on performance and adaptability in various scenarios.

Regarding your question on the architecture adaptation, yes, YOLOv5u adapts the YOLOv5 v6.1/6.0 architecture. The primary change lies in the detection head of the model, which has moved towards an anchor-free, objectness-free approach. This move enhances flexibility and detection efficiency, especially in complex scenarios where predefined anchor boxes might limit the detection capabilities.

If you're looking for more technical specifics or examples, I'd recommend diving into the [architecture description](https://docs.ultralytics.com/yolov5/tutorials/architecture_description/) you mentioned. While it focuses on an earlier version, the core architectural principles remain relevant, with the YOLOv5u modifications centered around the detection head for improved performance.

For direct experimentation or further exploration of the YOLOv5u's unique features, playing with the pre-trained models and reviewing their performance metrics might also offer valuable insights.

I hope this clarifies your queries! If you have more questions or need further assistance, feel free to ask. Happy detecting! 🚀

**Reply by MikolajBaranski** [reply]
Thank you that is a very helpful answer! Just to make a final clarification. I am using the YOLOv5u in my Masters Thesis and I would like to appropriately record the model I have used. For the YOLOv5xu.pt (the pretrained model), should I refer to it as the extra large version of the YOLOv5u v1.0, or is it the YOLOv5 v7.0, or is there some other versioning that would correctly identify this specific release of the model?

**Reply by pderrenger** [reply]
Hey, thanks for reaching out with your question! 😊

For your Masters Thesis, when documenting the model, it's best to refer to it as YOLOv5u, specifying the version as YOLOv5 v7.0 if that's the version you're utilizing, along with the specific variant you're using, for instance, `yolov5xu.pt` for the extra-large model. It's key to mention both the overall version (v7.0, as an example) and the specific model variant (`xu`) to provide clear context and specificity about the architecture and features employed in your work.

Here's how you could cite it:

"I utilized the YOLOv5u extra-large variant (yolov5xu.pt) from the YOLOv5 v7.0 release for object detection tasks in my project."

This approach ensures that your documentation is clear, precise, and tracks back to the correct version of the model relevant to your study.

Hope this helps, and all the best with your thesis! 🚀 If there's anything more you need, feel free to ask.

**Comment by Halafz7** [comment]
Good Content! I want to ask something. Is it possible to use yolov5u for Classification purposes?

**Reply by pderrenger** [reply]
@Halafz7 thank you for your kind words! 😊

YOLOv5u is primarily designed for object detection tasks and excels in that domain. However, if you're looking for a model specifically tailored for classification, you might want to consider using YOLOv8, which supports classification tasks directly.

Here's a quick example of how you can use YOLOv8 for classification:

### Python Example
```python
from ultralytics import YOLO

# Load a pretrained YOLOv8 classification model
model = YOLO("yolov8n-cls.pt")

# Run inference on an image
results = model("path/to/image.jpg")

# Display the results
results.show()
```

### CLI Example
```bash
# Run inference with a YOLOv8 classification model
yolo predict model=yolov8n-cls.pt source=path/to/image.jpg
```

For more details on using YOLOv8 for classification, you can refer to the [YOLOv8 documentation](https://docs.ultralytics.com/models/yolov8).

If you have any more questions or need further assistance, feel free to ask! 🚀

**Comment by chef-SAbhinay** [comment]
Can still use yolov5su.pt with YOLO from ultralytics pip package which is mentioned in the above docs?

**Reply by chef-SAbhinay** [reply]
or i have use torch.hub.load for it?

**Reply by chef-SAbhinay** [reply]
WARNING ⚠️ yolov5su.pt appears to require 'ultralytics.yolo', which is not in Ultralytics requirements.

**Reply by chef-SAbhinay** [reply]
i want to install model locally and use it

**Reply by glenn-jocher** [reply]
@Abhinay-Sadineni to use the `yolov5su.pt` model locally with the Ultralytics `YOLO` class, ensure you have the latest version of the `ultralytics` package installed. If you encounter issues, you can alternatively use `torch.hub.load` to load the model. For more details, visit the [YOLOv5 documentation](https://docs.ultralytics.com/models/yolov5/).

**Reply by chef-SAbhinay** [reply]
is this yolov5su.pt(yolov5 model) compatible with latest ultralytics package ?

my experiment setup doesn't allow the autoinstall 'ultralytics.yolo' (i mean external traffic is blocked)

@glenn-jocher 

will the latest package also issue autoinstall warning?

**Reply by chef-SAbhinay** [reply]
I believe torch.hub.load will also try to install model (which my experiment setup doesn't allow)

**Reply by chef-SAbhinay** [reply]
Ultralytics 8.2.81 is also giving the same error

**Reply by glenn-jocher** [reply]
To use `yolov5su.pt` with the latest Ultralytics package, ensure you have the latest version installed. If your setup blocks external traffic, you can manually download the model and dependencies. The autoinstall warning may still appear if dependencies are missing. For a seamless experience, consider using the model with the Ultralytics `YOLO` class after ensuring all dependencies are locally available.

**Reply by kanagarajan0709** [reply]
I built a custom dataset from Ultralytics Hub platform using Yolov5su model type and downloaded the trained model. When I tried to load getting this error : "An error occurred: 'Detect' object has no attribute 'grid'. "    

This is how I load the model : 
yolov5path - points to latest yolov5 git repo.
model path - refers to YOLOv5su model trained using Ultralytics Hub.

yolo_model = torch.hub.load(yolov5path, 'custom', force_reload=True, path=model_path, source='local')

YOLOv5  v7.0-338-gff063284 Python-3.12.4 torch-2.3.1+cpu CPU

YOLOv5s summary (fused): 193 layers, 9,114,632 parameters, 0 gradients, 23.8 GFLOPs
Adding AutoShape... 
An error occurred: 'Detect' object has no attribute 'grid'.

**Reply by glenn-jocher** [reply]
@kanagarajan0709 it appears that the `yolov5su.pt` model is not fully compatible with the latest Ultralytics package, leading to the "Detect object has no attribute 'grid'" error. To resolve this, please ensure you are using the correct version of the YOLOv5 repository that supports the `yolov5su.pt` model. If the issue persists, consider using the `torch.hub.load` method with the appropriate YOLOv5 version.

**Reply by kanagarajan0709** [reply]
>> To resolve this, please ensure you are using the correct version of the YOLOv5 repository that supports the yolov5su.pt model. If the issue persists, consider using the torch.hub.load method with the appropriate YOLOv5 version.

Yes ,  that's what I would be interested to know about. Could you please let me know which Yolov5 git hub repo that supports loading of yolov5su model. 

I am currently on latest git version of this - https://github.com/ultralytics/yolov5. 
Last commit e8a30cf820b273afbbf1c73a3f1088471a84fade (HEAD -> master, origin/master, origin/HEAD)
Author: Glenn Jocher <glenn.jocher@ultralytics.com>

Also I did use torch.hub.load method to load the model as below ,
yolo_model = torch.hub.load(yolov5path, 'custom', force_reload=True, path=model_path, source='local')

**Reply by glenn-jocher** [reply]
To load the `yolov5su.pt` model, please use the YOLOv5 repository from the `ultralytics/yolov5` GitHub. Ensure you are using a compatible version, as the latest updates might not support this model. If issues persist, consider using a specific commit that aligns with the model's requirements.

**Comment by MadhavKrishna1** [comment]
I would like to know about the training process of the YOLOv5 model. Specifically, is it trained on the entire COCO dataset?
I am planning to fine-tune a downloaded YOLOv5 model using additional images that are similar to those in the COCO dataset. These images have 80 classes and similar sizes to those in COCO
Could you provide guidance on finding this additional dataset?

Thank you!

**Reply by glenn-jocher** [reply]
@MadhavKrishna1 yOLOv5 models are indeed trained on the full COCO dataset, which includes 80 classes. For additional datasets similar to COCO, you might explore open-source repositories or platforms like Kaggle. Fine-tuning with your specific images should enhance performance on your target data.

**Comment by MadhavKrishna1** [comment]
1.Does the kaggle dataset have a label file also?
2.How can i fine tune the model ,is there any built in library
through which i could fine tune it.

**Reply by glenn-jocher** [reply]
@MadhavKrishna1 1. Yes, Kaggle datasets often include label files, but it's best to check the specific dataset details on Kaggle.
2. You can fine-tune the model using the Ultralytics YOLO Python API, which provides built-in support for training and fine-tuning.

**Comment by donia444** [comment]
hey ,i am trying to train yolov5nu on a custom dataset that is specifically for drone detection ,every time i train using this model ,i notice that my model overfits and gives a very very high unreal accuracy , but when i train on a different model like yolov5n that problem seem to not be found ,any idea why is that happening

**Reply by glenn-jocher** [reply]
@donia444 it seems that your YOLOv5nu model might be overfitting due to its more complex architecture compared to YOLOv5n, which is simpler and might generalize better on your dataset. Consider using techniques like data augmentation, dropout, or early stopping to mitigate overfitting. You can find additional strategies in the [Ultralytics documentation](https://docs.ultralytics.com/yolov5/) to help address this issue.

**Comment by PriyaRanjanReddy** [comment]
Hi can we use yolov5nu.pt for training on my own dataset and then use the trained model for commercial use where I can build an application over it

**Reply by glenn-jocher** [reply]
@PriyaRanjanReddy yes, you can use YOLOv5nu.pt for training on your dataset. For commercial use, you'll need an Ultralytics Enterprise License, which allows proprietary deployment without AGPL-3.0's open-source requirements. Visit https://ultralytics.com/license for licensing details.

**Reply by PriyaRanjanReddy** [reply]
What if i wanted to use yolo11 yolo8 or anything what license should I purchase

**Reply by PriyaRanjanReddy** [reply]
and please let me know the cost of the license as well

**Reply by glenn-jocher** [reply]
@PriyaRanjanReddy all current YOLO models, including YOLO11 and YOLOv8, require an Enterprise License for commercial use—pricing varies based on your project's needs, so please visit our licensing page to discuss details.
